bot/exts/moderation/verification.py

```diff
@@ -519,12 +539,34 @@ async def on_member_join(self, member: discord.Member) -> None:
         if member.guild.id != constants.Guild.id:
             return  # Only listen for PyDis events
 
+        raw_member = await self.bot.http.get_member(member.guild.id, member.id)
+
+        # Only send the message to users going through our gating system
+        if raw_member["is_pending"]:
+            await self.member_gating_cache.set(raw_member.id, True)
+            return
+
         log.trace(f"Sending on join message to new member: {member.id}")
         try:
             await safe_dm(member.send(ON_JOIN_MESSAGE))
         except discord.HTTPException:
             log.exception("DM dispatch failed on unexpected error code")
 
+    @Cog.listener()
+    async def on_member_update(self, before: discord.Member, after: discord.Member) -> None:
+        """Check if we need to send a verification DM to a gated user."""
+        before_roles = [r.id for r in before.roles]
+        after_roles = [r.id for r in after.roles]
+
+        if constants.Roles.verified not in before_roles and constants.Roles.verified in after_roles:
+            if await self.member_gating_cache.get(after.id):
+                try:
+                    await safe_dm(after.send(ALTERNATE_VERIFIED_MESSAGE))
+                except discord.HTTPException:
+                    log.exception("DM dispatch failed on unexpected error code")
+                finally:
+                    self.member_gating_cache.pop(after.id)
```

Comment:
Not sure I love using pop as delete. just do cache.delete if you don't need it.

Reply:
This was changed in 880b936faf83d8fa3d7489e1f9eaab89b93af1b8, but yep.